### PR TITLE
Remove transitive dependencies from encryption-sdk/build.gradle

### PR DESCRIPTION
**Pull request by reta**
### Description
Remove org.bouncycastle:bcprov-jdk15to18: not needed and breaks the plugins

### Related Issues
Closes https://github.com/opensearch-project/security/issues/3309

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: libs/encryption-sdk/build.gradle
Comment (reta) [inline libs/encryption-sdk/build.gradle:28]: This dependency is not used by encryption SDK and should be only included by the components that need it, right now it breaks at least `security` plugin since it completely messes up the security policies.

**Comment by reta** [comment]
@peternied @cwperks @willyborankin please check it out

File: libs/encryption-sdk/build.gradle
Comment (reta) [reply libs/encryption-sdk/build.gradle:28]: @vikasvb90  fyi

**Review by peternied — APPROVED** [review]

**Review by cwperks — APPROVED** [review]

**Comment by andrross** [comment]
Does this dependency come transitively via the [aws-encryption-sdk-java](https://github.com/aws/aws-encryption-sdk-java/blob/61b6260e1df17507cd4cb01a8a7d08b05d8fd745/pom.xml#L69-L73) library?

**Comment by reta** [comment]
> Does this dependency come transitively via the [aws-encryption-sdk-java](https://github.com/aws/aws-encryption-sdk-java/blob/61b6260e1df17507cd4cb01a8a7d08b05d8fd745/pom.xml#L69-L73) library?

Nope, 

```
+--- project :libs:opensearch-encryption-sdk
|    +--- project :libs:opensearch-common
|    +--- org.slf4j:slf4j-api:1.7.36
|    +--- commons-logging:commons-logging:1.2
|    +--- com.amazonaws:aws-encryption-sdk-java:2.4.0
|    \--- org.apache.commons:commons-lang3:3.13.0
```

**Comment by willyborankin** [comment]
> Does this dependency come transitively via the [aws-encryption-sdk-java](https://github.com/aws/aws-encryption-sdk-java/blob/61b6260e1df17507cd4cb01a8a7d08b05d8fd745/pom.xml#L69-L73) library?

no it was included :-(

**Review by willyborankin — APPROVED** [review]

**Comment by cwperks** [comment]
@reta it turns out many plugins are having separate dependency issues as well. I'll provide links in a minute.

What do you think about removing all of 

```
implementation "org.slf4j:slf4j-api:${versions.slf4j}"
implementation 'commons-logging:commons-logging:1.2'

// Encryption
implementation "org.bouncycastle:bcprov-jdk15to18:${versions.bouncycastle}"
implementation "org.apache.commons:commons-lang3:${versions.commonslang}"
```

from `opensearch-encryption-sdk`

**Review by DarshitChanpura — APPROVED** [review]

**Comment by reta** [comment]
> @reta it turns out many plugins are having separate dependency issues as well. I'll provide links in a minute.
> 
> What do you think about removing all of
> 
> ```
> implementation "org.slf4j:slf4j-api:${versions.slf4j}"
> implementation 'commons-logging:commons-logging:1.2'
> 
> // Encryption
> implementation "org.bouncycastle:bcprov-jdk15to18:${versions.bouncycastle}"
> implementation "org.apache.commons:commons-lang3:${versions.commonslang}"
> ```
> 
> from `opensearch-encryption-sdk`
> 
>     * Security analytics: https://github.com/opensearch-project/security-analytics/actions/runs/6081427926/job/16497056120?pr=533
> 
>     * Alerting: https://github.com/opensearch-project/alerting/actions/runs/6088707486/job/16519939511?pr=1114
> 
>     * ML Commons: https://build.ci.opensearch.org/blue/organizations/jenkins/distribution-build-opensearch/detail/distribution-build-opensearch/8372/pipeline/75/

Yeah, I also suspected that, please share the links, we will fix it

**Comment by cwperks** [comment]
@reta here are 3 examples. I'll update with more links if I find other plugins affected:

- Security analytics: https://github.com/opensearch-project/security-analytics/actions/runs/6081427926/job/16497056120?pr=533
- Alerting: https://github.com/opensearch-project/alerting/actions/runs/6088707486/job/16519939511?pr=1114
- ML Commons: https://build.ci.opensearch.org/blue/organizations/jenkins/distribution-build-opensearch/detail/distribution-build-opensearch/8372/pipeline/75/

File: libs/encryption-sdk/build.gradle
Comment (reta) [inline libs/encryption-sdk/build.gradle:23]: @cwperks I think we should be all set, thank you for heads up :heart:

File: libs/encryption-sdk/build.gradle
Comment (cwperks) [inline libs/encryption-sdk/build.gradle:24]: I don't see any imports from commons-lang3 in this library. Is this something that can be removed too or is it needed?

Comment (reta) [reply libs/encryption-sdk/build.gradle:24]: That's the transitive one for `aws` :( 

```
compileClasspath - Compile classpath for source set 'main'.
+--- project :libs:opensearch-common
+--- com.google.code.findbugs:annotations:3.0.1
+--- com.amazonaws:aws-encryption-sdk-java:2.4.0
\--- org.apache.commons:commons-lang3:3.13.0
```

**Review by peternied — APPROVED** [review]

**Comment by vikasvb90** [comment]
@reta We need to revert this change. BC is required by AWS encryption SDK which is currently not being used as we are waiting for approval from crypto BR team for custom changes. We have currently added just a NoOpCryptoHandler in place of it.

**Comment by reta** [comment]
> @reta We need to revert this change. BC is required by AWS encryption SDK which is currently not being used as we are waiting for approval from crypto BR team for custom changes. We have currently added just a NoOpCryptoHandler in place of it.

@vikasvb90  We cannot revert this change, the BC should be added to the component which uses it: I suspect this a plugin, right? Adding hard dependency on BC breaks the whole plugin ecosystem.

**Comment by reta** [comment]
@vikasvb90 additionally, please check the dependencies https://mvnrepository.com/artifact/com.amazonaws/aws-encryption-sdk-java/2.4.1: it potentially needs `bcprov-ext` but not `bcprov`

**Comment by vikasvb90** [comment]
@reta No, it is not the plugin which uses it but the lib itself. Let me check the bcprov dependency but we anyways need other dependencies and bcprov-ext dependency. Also, I am not sure if adding bcprov-ext can avoid changes required for security plugin.

**Comment by reta** [comment]
> @reta No, it is not the plugin which uses it but the lib itself. Let me check the bcprov dependency but we anyways need other dependencies and bcprov-ext dependency.

If lib uses it in any reasonable manner,  it should either:
 - fail to compile
 - fail at runtime (I would imply tests here)

Nothing from the above is happening. Please provide more details where it is needed because everything indicates the otherwise.

**Comment by reta** [comment]
> but we anyways need other dependencies and 

@vikasvb90 which other dependencies? the only transitive ones are common-lang3 and it has been kept intact. The library must not bring the commons-logging or slf4j-api (it could indicated it needs it) - this is the responsibility of end consumer to provide, either plugin / server / client / extension.

**Comment by vikasvb90** [comment]
You are right that it is not being used right now. As I mentioned earlier, that this had changes done in encryption SDK to support partial encryption/decryption which is yet to undergo a review from SDK team itself and that's why it is just a NoOpCryptoHandler. Let me get back on how we can enforce its usage in lib meanwhile but inclusion of the dependency is something we need.

File: libs/encryption-sdk/build.gradle
Comment (navneet1v) [reply libs/encryption-sdk/build.gradle:24]: @reta But this causing downstream plugins to break. :( https://github.com/opensearch-project/neural-search/actions/runs/6094181033/job/16535187721?pr=287

Comment (reta) [reply libs/encryption-sdk/build.gradle:24]: @navneet1v I know :( but we a) either completely drop the  `encryption-sdk` or b) the `neural-search` could just exclude the common-lang3 from dependencies, the sql plugin has same issue

Comment (vikasvb90) [reply libs/encryption-sdk/build.gradle:24]: @navneet1v This should get auto-fixed once https://github.com/opensearch-project/OpenSearch/pull/9833 or https://github.com/opensearch-project/OpenSearch/pull/9812 is merged

Comment (navneet1v) [reply libs/encryption-sdk/build.gradle:24]: @vikasvb90 thanks :)

**Pull request merged**
